lucet-runtime/lucet-runtime-internals/src/instance/signals.rs

```diff
@@ -245,12 +247,29 @@ extern "C" fn handle_signal(signum: c_int, siginfo_ptr: *mut siginfo_t, ucontext
                 true
             }
         };
-
+        // If we are switching to the host context and we are not processing a SIGALRM from a
+        // KillSwitch, then we were either externally signalled in some other manner, or we faulted
+        // during execution in a guest region.
+        //
+        // In either case, we will need to unlock and update the execution domain ourselves.
         if switch_to_host {
-            // we must disable termination so no KillSwitch may fire in host code.
-            inst.kill_state.disable_termination();
+            // This debug assert exists to make sure that this is not a blocking operation.
+            debug_assert_eq!(inst.kill_state.domain_is_locked(), false);
+            let current_domain = inst.kill_state.execution_domain();
+            match current_domain {
+                Ok(mut current_domain) => {
+                    // If we are here, we should still be in the guest domain.
+                    debug_assert_eq!(*current_domain, execution::Domain::Guest);
+                    *current_domain = execution::Domain::Terminated;
+                }
+                Err(poison_error) => {
+                    // Because a signal must never be sent while the host context is active, we
+                    // will update the domain even if the mutex has been poisoned.
+                    let mut current_domain = poison_error.into_inner();
+                    *current_domain = execution::Domain::Terminated;
+                }
```

Comment:
So :point_up:  this whole block is one of the places where pulling `terminable` out _does_ make things more complicated. Given that this is inside of a signal handler, what should we do here?

Reply:
We don't need it! :grin: There's a comment above re. `domain_is_locked`, but there's another concern this introduces: we could be handling a to-be-fatal `SIGSEGV` right at this point having grabbed the lock when the guest times out. Then, the `SIGALRM` causes us to re-enter this function, would also `switch_to_host`, reach here, attempt to grab the lock again, and deadlock.

I think the signal handler will be fine in not updating `execution_domain`, and if we really want to set it to `Terminated`, we should do so in `KillSwitch` once the `tid_change_notifier` has done its thing. Even if we don't, a second `KillSwitch` wouldn't signal, because the shared `thread_id` would have been updated to `None`, but I think reiterating that the referenced instance is terminated is good for clarity.